### PR TITLE
Positional bool

### LINKED ISSUE #171: Unable to derive `positional::<bool>` fields
I want to create a positional that can take boolean values, i.e. literally `true`, and `false`, no fancy parsing necessary.

I started off with this to parse a number in a related command:

```rust
#[derive(Bpaf)]
struct Args {
    #[bpaf(positional("a number"))]
    argument: u32
}
```

changing the `u32` to bool however does not seem to be possible

```rust
#[derive(Bpaf)]
struct Args {
    #[bpaf(positional("a bool"))]
    argument: bool
}
```

results in a compiler error

```
error: Can't derive consumer for unnamed boolean field
  --> crates/---/src/main.rs:71:15
   |
71 |     argument: bool
   |               ^^^^
```

approaches of setting a different parser similarly failed.

```rust
#[bpaf(positional("a bool"), parser(FromStr:from_str)]
#[bpaf(positional("a bool"), parser(bool::parse)]
```

the only thing i found to work is to manually create a parser for the field which is a bit unhandy and inconsistent:

```rust
#[derive(Bpaf)]
struct Args {
    #[bpaf(external(parse_bool)]
    argument: bool
}

fn parse_bool() -> impl Parser<bool> {
	positional("a bool")
}
```

### PR DESCRIPTION
**Pull request by pacak**
Fixes #171

@ysndr, can you take a look if this does what you want? If it is - I'll make a new release.

Todo:
- [x] Add `anywhere` top level annotation
- [x] Add `box` top level annotation
- [x] Add `catch` for `anywhere` and use it in failing test cases
- [x] update documentation

### CODE REVIEW

**Comment by ysndr** [comment]
seems so, found a different issue though:

remember that i try to parse `--setBool <key> <value>`?

```rust
#[derive(Debug, Clone, Bpaf)]
#[bpaf(adjacent)]
#[allow(unused)]
pub struct ConfigSetBool {
    /// Set <key> to <bool>
    #[bpaf(long("setBool"))]
    set_bool: (),
    /// Configuration key
    #[bpaf(positional("key"))]
    key: String,
    /// Configuration Value (bool)
    #[bpaf(positional("bool"))] // << this seems to work now, hurray :)
    value: bool,
}
```

I use this ^ contraption for that.

if i run `xyz config --setBool key notabool` I'm greeted with 

```
ERROR: --setBool is not expected in this context
    
    configure user parameters
```

Which yes, tells me somewhat that it cant parse `--setBool`. but the way it does does not really help me to do it right. -- what i would expect is it saying which (positional) argument failed to parse.

This also seems to happen with when `value: u32` but I'm not sure if that is a regression maybe

**Comment by pacak** [comment]
`adjacent` makes it so blocks must stick together, you also need `anywhere`

**Comment by pacak** [comment]
```rust
pub struct ConfigSetBool {
    /// Set <key> to <bool>
    #[bpaf(long("setBool"))]
    set_bool: (),
    /// Configuration key
    #[bpaf(positional("key"))]
    key: String,
    /// Configuration Value (bool)
    #[bpaf(positional("bool"))] // << this seems to work now, hurray :)
    value: bool,
}


fn try_this() -> impl Parser<ConfigSetBool> {
    config_set_bool().anywhere()
}
```

If this works - I'll add `anywhere` to a derive macro

**Comment by ysndr** [comment]
no, that doesnt change anything as far as i can tell

**Comment by ysndr** [comment]
Unrelated,... wouldnt it be great if i could write 
```
fn try_this<T: ToParser>() -> impl Parser<T> {
    T::to_parser().anywhere()
}
```

instead of the same function three times ;)

**Comment by pacak** [comment]
Hmm... It sort of works for me

```rust
use bpaf::*;

#[derive(Debug, Clone, Bpaf)]
#[bpaf(adjacent)]
#[allow(unused)]
pub struct ConfigSetBool {
    /// Set <key> to <bool>
    #[bpaf(long("setBool"))]
    set_bool: (),
    /// Configuration key
    #[bpaf(positional("key"))]
    key: String,
    /// Configuration Value (bool)
    #[bpaf(positional("bool"))] // << this seems to work now, hurray :)
    value: bool,
}

fn main() {
    let x = config_set_bool().anywhere().many().to_options().run();
    todo!("{:?}", x);
}
```

```console
elakelaiset% cargo run --release --example set_bool -- --setBool banana false --setBool durian true
    Finished release [optimized] target(s) in 0.02s
     Running `target/release/examples/set_bool --setBool banana false --setBool durian true`
thread 'main' panicked at 'not yet implemented: [ConfigSetBool { set_bool: (), key: "banana", value: false }, ConfigSetBool { set_bool: (), key: "durian", value: true }]', examples/set_bool.rs:21:5
```

**Comment by ysndr** [comment]
try `--setBool banana 123`

**Comment by pacak** [comment]
Hmm... I see, I would call this a bug in `anywhere`, it should ignore missing arguments but should retain parse errors. Will fix, might take a bit.

**Comment by pacak** [comment]
> instead of the same function three times ;)

Without functional dependencies or type families it seems hard to implement the trait constraint `ToParser`... I blame Rust :)

**Comment by ysndr** [comment]
> I blame Rust :)

fair enough :D

Though I'd be perfectly happy if deriving `Bpaf` for something would implement a trait for them that just exposes their derived parser as a `Box<dyn Parser>`..
Ive been doing that manually so far but having access to that directly would be handy in such cases...

**Comment by pacak** [comment]
> try --setBool banana 123, parsing correct values works for me too

Pushed something. Seems to work for me as expected with invalid values. I'll have to add some tests and possibly deal with more corner cases so release might take a bit.

**Comment by pacak** [comment]
> Though I'd be perfectly happy if deriving Bpaf for something would implement a trait for them that just exposes their derived parser as a Box<dyn Parser>..

Hmm... Currently you can get a boxed parser with `construct!(parser)`, I can also expose it as method and add the method to top level bpaf annotation...

```rust
#[derive(Debug, Clone, Bpaf)]
#[bpaf(adjacent, anywhere, box)]
struct Foo {
...
}
```

in this case will give you `fn foo() -> Box<dyn Parse>`...

**Comment by pacak** [comment]
Any feedback on the article?

Back in 40 minutes.

**Comment by ysndr** [comment]
well i kinda just want to have a generic way to get to the derived parser, what i currently use is a macro like this:

```rust
  macro_rules! parseable {
        ($type:ty, $parser:ident) => {
            impl crate::commands::package::Parseable for $type {
                fn parse() -> bpaf::parsers::ParseBox<Self> {
                    let p = $parser();
                    bpaf::construct!(p)
                }
            }
        };
    }
```

that i just run manually for every type where i would want that capability..
it gets me what i want but if the `Bpaf` derive macro could do that automatically would just be more convenient.

> Any feedback on the article? 

reading now

**Comment by ysndr** [comment]
> Any feedback on the article?

shall I review there, or do you want to discuss outside?

**Comment by pacak** [comment]
> shall I review there, or do you want to discuss outside?

Up to you, I'm available at discord, whatsapp, google chat and signal. Probably google chat is going to be the easiest if you use it. manpacket@gmail.com

**Comment by ysndr** [comment]
> google chat

Oh they try it again? :D

I'll reach out tomorrow I guess or commit the review comments first, its not quite the middle of the day for me, quite the opposite

Though it is interesting to see this project unravelled a bit 👍🏼

**Comment by ysndr** [comment]
> > Though I'd be perfectly happy if deriving Bpaf for something would implement a trait for them that just exposes their derived parser as a Box..
> 
> Hmm... Currently you can get a boxed parser with `construct!(parser)`, I can also expose it as method and add the method to top level bpaf annotation...
> 
> ```rust
> #[derive(Debug, Clone, Bpaf)]
> #[bpaf(adjacent, anywhere, box)]
> struct Foo {
> ...
> }
> ```
> 
> in this case will give you `fn foo() -> Box<dyn Parse>`...

Oh in don't know if were talking of different things then..

I mean something like `<Foo as AsParseBox>::parse_box() -> Box<dyn Parser>`

**Comment by pacak** [comment]
related to https://github.com/pacak/bpaf/pull/170, contains breaking changes in behavior

**Comment by ysndr** [comment]
Sorry this will be a bit of a brain dump i figured after writing this up.

tl;dr

```
#[positional]
filed: bool
```

seems to work, error messages now talk about the value part not something else.
Nice progress!

Errors are something else, give it a read, I'll make it a seperate issue not to block this one.

So with

``` rust
#[derive(Bpaf, Debug)]
struct Xyz {

    #[bpaf(long, switch)]
    bool_flag: bool,

    #[bpaf(long, argument)]
    bool_opt: bool,

    #[bpaf(positional)]
    bool_arg: bool,
}
```


these **work** as expected:

```
$ cargo run --  --bool-opt true true  --bool-flag 
$ cargo run --  --bool-opt true --bool-flag true
$ cargo run --  --bool-opt true true 
```

these **fail** as expected:

```
$ cargo run --  --bool-opt true nottrue 
Couldn't parse "nottrue": provided string was not `true` or `false`
               ^       ^                          ^    ^    ^     ^
               '       '                          '----'----'-----'
               '-------'- why are these not the same? -'
``` 


with 

```rust 
#[derive(Debug, Clone, Bpaf)]
#[bpaf(adjacent)]
#[allow(unused)]
pub struct ConfigSetBool {
    /// Set <key> to <bool>
    #[bpaf(long("setBool"))]
    set_bool: (),
    /// Configuration key
    #[bpaf(positional("key"))]
    key: String,
    /// Configuration Value (bool)
    #[bpaf(positional("bool"))] // << this seems to work now, hurray :)
    value: bool,
}
```

fails correctly

```
$ cargo run --  --setBool key tru
Couldn't parse "tru": provided string was not `true` or `false`
```


`boxed` seems to work for what i can tell, even though it might not be what i proposed hh


<details>
<summary>UX discussion</summary>

I think this better go into a separate issue but i put it here for context


```
$ cargo run -- --bool-opt tru tru
Couldn't parse "tru": provided string was not `true` or `false`
                 |    L use positive sentence instead
                  \ which tru (see below)?

$ cargo run -- --bool-opt true true --bool-fla
No such flag: `--bool-fla`, did you mean `--bool-flag`?
            L differnet        


$ cargo run -- true  --bool-flag --bool-flag --bool-opt true true
--bool-flag is not expected in this context
|
 \ another one (this time without ticks at all :p)

```  

the error message should if possible show which argument went wrong

```   
$ cargo run -- --bool-opt tru true                
Invalid argument:

       got: --bool-opt tru
  expected: --bool-opt (true | false)
   
Type '<command> <subcommand> --help' to see all arguments
```

```   
$ cargo run -- --bool-opt true true --bool-fla              
No such flag:  

       got: --bool-fla
  expected: [ --bool-flag ] < Only if --bool-flag is not already specified
   
Type '<command> <subcommand> --help' to see all arguments
```

```
cargo run -- true  --bool-flag --bool-flag --bool-opt true true
No such flag:  

       got: --bool-flag
  expected: (none)
   
Type '<command> <subcommand> --help' to see all arguments

or even better:

$ cargo run -- true  --bool-flag --bool-flag --bool-opt true true
Duplicate flag:  

       got: --bool-flag --bool-flag
  expected: [ --bool-flag ]
   
Type '<command> <subcommand> --help' to see all arguments
```


Though using positionals  for multi value options is still a bit awkward and _looks_ like a hack in the help:

```
Usage: [--bool-flag] --bool-opt ARG <ARG> --setBool <key> <bool>

Available positional items:
    <key>   Configuration key          <---,
    <bool>  Configuration Value (bool) <---,
                                           ,
                                           ,
Available options:                         ,
        --bool-flag                        ,
        --bool-opt <ARG>                <--
```


</details>

**Comment by pacak** [comment]
FWIW pushed a branch that renders proper help for multivariable arguments - `multiarg`

**Comment by pacak** [comment]
Positional `bool` is now supported, 
For purely positional items something like this should work in the recent master, `adjacent` is not needed. I also made some changes to make error messages more user friendly

```rust
#[derive(Debug, Clone, Bpaf)]
#[bpaf(anywhere, box)]
struct Foo {
    /// Set <key> to <bool>
    #[bpaf(long("setBool"))]
    set_bool: (),
    /// Configuration key
    #[bpaf(positional("key"))]
    key: String,
    /// Configuration Value (bool)
    #[bpaf(positional("bool"))] // << this seems to work now, hurray :)
    value: bool,
}
```

Took out UX discussion parts in a separate issue. going to merge this.
